### PR TITLE
Implement Filter

### PR DESCRIPTION
**Pull request by adamgfraser**

### CODE REVIEW

File: core/shared/src/main/scala/zio/prelude/AssociativeFlatten.scala [hunk 1]
```diff
@@ -240,4 +240,73 @@ trait AssociativeFlattenSyntax {
...
+      covariant: Covariant[F],
+      identityBoth: IdentityBoth[F],
+      identityEither: IdentityEither[F]
+    ): F[A] =
+      fa.flatMap(a => f(a).flatMap(b => if (b) a.succeed else identityEither.none))
+
+    /**
+     * Filters an `F[A]` value with a predicate `f`.
+     */
+    def filterNot[B](f: A => Boolean)(implicit
```
Comment (sideeffffect) [inline core/shared/src/main/scala/zio/prelude/AssociativeFlatten.scala]: Is it really worth having?

File: core/shared/src/main/scala/zio/prelude/AssociativeFlatten.scala [hunk 2]
```diff
@@ -240,4 +240,73 @@ trait AssociativeFlattenSyntax {
...
+      flatten: AssociativeFlatten[F],
+      covariant: Covariant[F],
+      identityBoth: IdentityBoth[F],
+      identityEither: IdentityEither[F]
+    ): F[A] =
+      filter(a => !f(a))
+
+    def withFilter(f: A => Boolean)(implicit
```
Comment (sideeffffect) [inline core/shared/src/main/scala/zio/prelude/AssociativeFlatten.scala:295]: Why have both `filter` and `withFilter`?

**Review by sideeffffect — DISMISSED** [review]
Looks good, I just 2 questions.

File: core/shared/src/main/scala/zio/prelude/AssociativeFlatten.scala (hunk 1 above)
Comment (adamgfraser) [reply core/shared/src/main/scala/zio/prelude/AssociativeFlatten.scala]: It is definitely not fundamental. I included it for consistently with the Scala standard library but we could take it out.

File: core/shared/src/main/scala/zio/prelude/AssociativeFlatten.scala (hunk 2 above)
Comment (adamgfraser) [reply core/shared/src/main/scala/zio/prelude/AssociativeFlatten.scala:295]: For for comprehensions.

Comment (sideeffffect) [reply core/shared/src/main/scala/zio/prelude/AssociativeFlatten.scala:295]: Aha! I see, neat.
Do you think you could put that to a comment? I think I'm not the only one who wouldn't immediately think of for-comprehensions.

File: core/shared/src/main/scala/zio/prelude/AssociativeFlatten.scala (hunk 1 above)
Comment (sideeffffect) [reply core/shared/src/main/scala/zio/prelude/AssociativeFlatten.scala]: I don't mind keeping it. But I have to admit I have always wondered about the value of these `*Not` variants even in the standard library.

**Review by sideeffffect — DISMISSED** [review]
Adding a comment with explanation would be great.

File: core/shared/src/main/scala/zio/prelude/AssociativeFlatten.scala [hunk 3]
```diff
@@ -240,4 +240,73 @@ trait AssociativeFlattenSyntax {
     def flatMap[B](f: A => F[B])(implicit flatten: AssociativeFlatten[F], covariant: Covariant[F]): F[B] =
       flatten.flatten(covariant.map(f)(fa))
   }
+
+  /**
+   * Provides infix syntax for filtering covariant types.
+   */
+  implicit class AssociativeFlattenCovariantIdentityBothIdentityEitherOps[F[+_], A](fa: F[A]) {
```
Comment (0pg) [inline core/shared/src/main/scala/zio/prelude/AssociativeFlatten.scala:247]: this class is composition of several typeclasses, but why do you add this class in here?
I would like to know if there are any rules for this.

File: core/shared/src/main/scala/zio/prelude/AssociativeFlatten.scala [hunk 4]
```diff
@@ -240,4 +240,62 @@ trait AssociativeFlattenSyntax {
...
+    ): F[B] =
+      fa.flatMap(a => pf.lift(a).getOrElse(identityEither.none))
+
+    /**
+     * Filters an `F[A]` value with a predicate `f`.
+     */
+    def filter[B](f: A => Boolean)(implicit
+      flatten: AssociativeFlatten[F],
+      covariant: Covariant[F],
+      identityBoth: IdentityBoth[F],
```
Comment (sideeffffect) [inline core/shared/src/main/scala/zio/prelude/AssociativeFlatten.scala:276]: Do we really need `IdentityBoth` here?

File: core/shared/src/main/scala/zio/prelude/AssociativeFlatten.scala (hunk 1 above)
Comment (adamgfraser) [reply core/shared/src/main/scala/zio/prelude/AssociativeFlatten.scala]: I took it out.

File: core/shared/src/main/scala/zio/prelude/AssociativeFlatten.scala (hunk 4 above)
Comment (adamgfraser) [reply core/shared/src/main/scala/zio/prelude/AssociativeFlatten.scala:276]: Yes we can relax this constraint!

File: core/shared/src/main/scala/zio/prelude/AssociativeFlatten.scala (hunk 2 above)
Comment (adamgfraser) [reply core/shared/src/main/scala/zio/prelude/AssociativeFlatten.scala:295]: Done!

File: core/shared/src/main/scala/zio/prelude/AssociativeFlatten.scala (hunk 4 above)
Comment (sideeffffect) [reply core/shared/src/main/scala/zio/prelude/AssociativeFlatten.scala:276]: I still see it here :smile: 
Btw, can you also double check other methods, that they don't contain unnecessary implicits?

Comment (adamgfraser) [reply core/shared/src/main/scala/zio/prelude/AssociativeFlatten.scala:276]: We need it for `succeed.

**Review by sideeffffect — APPROVED** [review]
Thanks, Adam! :+1:

**Pull request merged**